src/mcts/search.cc

```diff
@@ -1145,6 +1145,20 @@ SearchWorker::NodeToProcess SearchWorker::PickNodeToExtend(
       }
       return NodeToProcess::Collision(node, depth, collision_limit);
     }
+    // Probably best place to check for two-fold draws consistently.
+    // Depth starts with 1 at root, so real depth is depth - 1.
+    if (node->IsTwoFoldTerminal() && params_.GetTwoFoldDrawLevel() > 0) {
+      LOGFILE << "== encountered twofold draw == depth: " << depth - 1;
+      if ((params_.GetTwoFoldDrawLevel() == 2 && depth - 1 <= 3) ||
+        // Level 2: no two-fold draw at depth 3 or lower
+          (params_.GetTwoFoldDrawLevel() == 1 && depth - 1 < node->GetM())) {
+        // Level 1: check whether first repetition was before root
+        // Length of repetition was stored in m_.
+        // TODO: Cleaner alternative to call position history?
+        LOGFILE << "== twofold draw reset ==";
+        node->MakeNotTerminal();
+      }
+    }
     // Either terminal or unexamined leaf node -- the end of this playout.
     if (node->IsTerminal() || !node->HasChildren()) {
```

Comment:
If you notice a performance issue, you can try moving the two fold not terminal check under terminal check as a node can only be two fold if a terminal.

Reply:
The idea behind making this into a separate statement was that when a twofold terminal is made non terminal through tree reuse, we already want the first visit to it be processed as a "regular" visit on that node, so we mark it non terminal on the fly and continue then.

Reply:
Yes, I realize the latter branch still wants to check if it's still a terminal, but this is one of the hottest loops in search, and checking if every node is a two fold involves loading the bitfield then masking and checking for the specific enum value whereas the terminal check is slightly faster because it's the common case default value 0. But might not need optimizing in the first place.

Reply:
The following seems marginally faster. Note that I also removed the unnecessary `params_.GetTwoFoldDraws()` call.
```diff
diff --git a/src/mcts/search.cc b/src/mcts/search.cc
index 224d080..6c0ea35 100644
--- a/src/mcts/search.cc
+++ b/src/mcts/search.cc
@@ -1145,21 +1145,20 @@ SearchWorker::NodeToProcess SearchWorker::PickNodeToExtend(
       }
       return NodeToProcess::Collision(node, depth, collision_limit);
     }
-    // Probably best place to check for two-fold draws consistently.
-    // Depth starts with 1 at root, so real depth is depth - 1.
-    if (node->IsTwoFoldTerminal() && params_.GetTwoFoldDraws()) {
-      if (depth - 1 < node->GetM()) {
+    // Either terminal or unexamined leaf node -- the end of this playout.
+    if (node->IsTerminal() || !node->HasChildren()) {
+      // Probably best place to check for two-fold draws consistently.
+      // Depth starts with 1 at root, so real depth is depth - 1.
+      if (node->IsTwoFoldTerminal() && depth - 1 < node->GetM()) {
         // Check whether first repetition was before root. If yes, remove
         // terminal status of node.
         // Length of repetition was stored in m_.
         // TODO: Cleaner alternative to call position history?
         node->MakeNotTerminal();
+      } else {
+        return NodeToProcess::Visit(node, depth);
       }
     }
-    // Either terminal or unexamined leaf node -- the end of this playout.
-    if (node->IsTerminal() || !node->HasChildren()) {
-      return NodeToProcess::Visit(node, depth);
-    }
     Node* possible_shortcut_child = node->GetCachedBestChild();
     if (possible_shortcut_child) {
       // Add two here to reverse the conservatism that goes into calculating the
```

Reply:
Yeah, I just blindly replaced all `Level` calls by the new boolean, not realizing that in this case, the existence of a twofold terminal implies that we in fact work with twofold terminals. Will implement these suggested changes.

Reply:
I think I have to understand the logic again here. Does a move which isn't a twofold terminal draw anymore have children or not? I hope to be correct in what we expect here is that we can expand it in the same visit which then gets that node already its first visit and explores a child.

Reply:
I came to the conclusion that this reordering doesn't handle it correctly if a terminal twofold after being made non terminal has no edges (which it doesn't); logic should return that node for expanding, but it doesn't.

Reply:
Right, this case was indeed wrong.

Reply:
We could still move it inside the `IsTerminal()` condition, and add another `return NodeToProcess::Visit(node, depth)` after checking for no children; not sure whether that's cheaper than current double checking.

Reply:
I tried a different approach, moving the `!node->HasChildren()` in a separate if statement, as I think this is the cleaner alternative, but I didn't like the added complexity.

Reply:
Changed it the way you describe. It now fits the logic better: If a node is terminal, we check whether it should stay terminal, and either make it non terminal (twofold draw after playing into the variation and tree reuse) or return `NodeToProcess::Visit(node, depth)`. Anything which survives the statement isn't a terminal, and we continue with the regular checks.

Reply:
In theory, a twofold terminal which just got reverted won't have children anyway, so we could return the `NodeToProcess` directly without checking for children. However, this case will be so rare that the additional code can't be considered to be an optimization, only a code obfuscation.